sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,360 @@
+        ILogger logger = loggingOptions.LoggerFactory.CreateLogger<ClientLoggingPolicy>();
+        _handler = new LoggingHandler(logger, sanitizer);
+    }
+
+    /// <inheritdoc/>
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex) =>
+        ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+
+    /// <inheritdoc/>
+    public override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex) =>
+        await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        if (!_handler.IsEnabled(LogLevel.Warning, EventLevel.Warning)) // Warning is the highest level logged
+        {
+            if (async)
+            {
+                await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline, currentIndex);
+            }
+            return;
+        }
+
+        PipelineRequest request = message.Request;
+
+        string requestId = Guid.NewGuid().ToString();
+        message.LoggingCorrelationId = requestId;
+
+        _handler.LogRequest(request, requestId, _clientAssembly);
+
+        byte[]? bytes = null;
+        Encoding? requestTextEncoding = null;
+
+        if (_logContent && request.Content != null && _handler.IsEnabled(LogLevel.Information, EventLevel.Informational))
+        {
+            // Convert binary content to bytes
+            using var memoryStream = new MaxLengthStream(_maxLength);
+            if (async)
+            {
+                await request.Content.WriteToAsync(memoryStream, message.CancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                request.Content.WriteTo(memoryStream, message.CancellationToken);
+            }
+            bytes = memoryStream.ToArray();
+
+            // Try to extract a text encoding from the headers
+            if (request.Headers.TryGetValue("Content-Type", out var contentType) && contentType != null)
+            {
+                ContentTypeUtilities.TryGetTextEncoding(contentType, out requestTextEncoding);
+            }
+
+            _handler.LogRequestContent(requestId, bytes, requestTextEncoding);
+        }
+
+        var before = Stopwatch.GetTimestamp();
+
+        try
+        {
+            if (async)
+            {
+                await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline, currentIndex);
+            }
+        }
+        catch (Exception ex)
+        {
+            _handler.LogExceptionResponse(requestId, ex);
```

Comment:
Is the exception independent of the HTTP request/response?  If so, would it be possible to move this event so that it is logged in the transport instead?

The reason for this is that if an end-user wanted to replace HTTP logging with a different policy or use HttpClient logging instead, we would want to still have a way to log exceptions.

Reply:
Hmm well Azure.Core considers this log as an informational HTTP request/response log called "Exception response" - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Diagnostics/AzureCoreEventSource.cs#L259 to me I think this belongs as part of the message logging. I don't think I should change the event level in case we end up using the event source for backwards compatibility in Azure.Core.

Reply:
So it's not really an exception log in the typical sense, it's more of an informational control flow log related to message logging.

Reply:
What exceptions could be thrown here that would be logged?

Reply:
Another way of thinking about this one is ... if someone wanted to use HttpClient logging for HTTP requests and responses, would we want them to stop logging this exception?

Reply:
This would be the case that the transport throws an exception. I would assume HttpClient would log that as part of their request/response logging.

Reply:
The transport, or the runtime, or the HttpClient, or any policies that are invoked by the call to `ProcessNext`.

I'm concerned about expecting HttpClient to handle this, because HttpClient logging may not do this, may not be turned on by default, etc.  One of our feature requirements is that users can collect logs to investigate errors without needing to make changes to the app.  If Aspire injected an HttpClient to log HTTP traffic but didn't log e.g. OutOfMemoryException, this would turn off all exception logging for the client.  I don't think that is an outcome we want.

Reply:
If we're getting OOM exceptions, the process is most likely dying, and we can't safely log.  Not sure that I'm following the scenario in the previous comment.

Reply:
I ended up moving this one